harness/determined/pytorch/_pytorch_context.py

```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+                f"Enabling mixed precision training with opt_level: {self.mixed_precision}."
+            )
+            model, optimizers = apex.amp.initialize(
+                model,
+                optimizers,
+                opt_level=self.mixed_precision,
+                verbosity=1
+                if self.distributed.get_rank() == 0 or self.env.experiment_config.debug_enabled()
+                else 0,
+            )
+            self.model = model
+            if not isinstance(optimizers, list):
+                self.optimizers = [optimizers]
+        return model, optimizers
+
+    def _should_communicate_and_update(self) -> bool:
+        if self.current_batch_idx is None:
+            raise det.errors.InternalException("Training hasn't started.")
+        return (self.current_batch_idx + 1) % self.hvd_config.aggregation_frequency == 0
+
+    def _backward(self, loss: torch.Tensor):  # type: ignore
+        if self._use_amp():
+            with apex.amp.scale_loss(loss, self.optimizers) as scaled_loss:
+                scaled_loss.backward()
+
+                if self.hvd_config.use and self._should_communicate_and_update():
+                    # When using horovod, we need to finish communicating gradient
+                    # updates before they are unscaled which happens when we exit
+                    # of this context manager.
+                    for optimizer in self.optimizers:
+                        optimizer.synchronize()
+        else:
+            loss.backward()
+
+            # Communication needs to be synchronized so that is completed
+            # before we apply gradient clipping and `step()`.
+            if self.hvd_config.use and self._should_communicate_and_update():
+                for optimizer in self.optimizers:
+                    optimizer.synchronize()
+
+    def _average_gradients(self, parameters: Any, divisor: int) -> None:
+        check.gt_eq(divisor, 1)
+        if divisor == 1:
+            return
+
+        divisor_value = float(divisor)
+        for p in filter(lambda param: param.grad is not None, parameters):
+            p.grad.data.div_(divisor_value)
+
+    def _step_optimizer(self, optimizer: torch.optim.Optimizer) -> None:  # type: ignore
+        # Communication needs to be synchronized so that is completed
+        # before we apply gradient clipping and `step()`.
+        if self._should_communicate_and_update():
+            parameters = (
+                [p for group in optimizer.param_groups for p in group.get("params", [])]
+                if not self._use_amp()
+                else apex.amp.master_params(optimizer)
+            )
+
+            if self.hvd_config.average_aggregated_gradients:
+                self._average_gradients(
+                    parameters=parameters, divisor=self.hvd_config.aggregation_frequency
+                )
+
+            for callback in self.callbacks.values():
+                callback.on_before_optimizer_step(parameters)
```

Comment:
question: are we going to keep this callback in the final API?

Reply:
Want to remove this in the final API. Added a comment for future removing.